openedx/features/wikimedia_features/meta_translations/utils.py

```diff
@@ -0,0 +1,263 @@
+                    existing_data.data_type, existing_data.data, outline_block_data
+                ))
+                existing_data.data = outline_block_data
+                existing_data.save()
+    except CourseBlock.DoesNotExist:
+        # Add new blocks in db for new modules/components in course outline.
+        log.info("Add base course block.")
+        CourseBlock.create_course_block_from_dict(outline_block_dict, course_key)
+
+
+def map_translated_course_block(existing_course_blocks, outline_block_dict, course_key, base_course_blocks_data):
+    """
+    Map translated-course block -> Sync Course Outline block to CourseBLock table and crete Translation
+    Mapping entries by comparing data of base-course and translated course block.
+
+    - if block does not exist, create new course-block and translation mapping by comparing
+      block-data with base-course data.
+    - if data exists, check for existing translation mapping and set overwrite True if user
+      has changed data from front-end.
+
+
+    Arguments:
+        existing_course_blocks (CourseBlock): existing course-blocks in db for given course-key
+        outline_block_dict (dict): contains data of course-outline block
+        course_key (CourseKey): translated course version on which mapping needs to perform
+        base_course_blocks_data: source/base course blocks data so that translation mapping can be created.
+    """
+    try:
+        course_block = existing_course_blocks.get(block_id=outline_block_dict.get("id"))
+        for key, value in outline_block_dict.get("data", {}).items():
+            try:
+                wiki_translation = WikiTranslation.objects.get(source_block_data__data_type=key, target_block=course_block)
+
+                # if re-run outline data is neither equal to original string (source block data) nor equal to translated string
+                # check value of applied, if applied is also True that means user might have updated data from front-end.
+                if value != wiki_translation.translation and value != wiki_translation.source_block_data.data:
+                    log.info("Mapping found, but data is not same as original and translated string for block {}".format(
+                        json.dumps(outline_block_dict))
+                    )
+                    log.info("Source data: {}".format(value))
+                    log.info("Translated data: {}".format(wiki_translation.translation))
+                    if wiki_translation.applied:
+                        log.info("Setting overwrite True.")
+                        wiki_translation.overwrite = True
+                        wiki_translation.save()
+                    else:
+                        # if applied is False, this means that latest translation is not applied yet.
+                        log.info("Latest translation is not applied yet.")
+
+            except WikiTranslation.DoesNotExist:
```

Comment:
same here, when it is supposed to occur?

Reply:
- user creates base course with single section
-  user creates re run
- user creates mapping -> that will create translation objects.
- user deletes section from source-course, now mapping will be lost as source course related data is deleted and hence translation object will also be deleted.

Reply:
can we move this except block right after try block,

```
try:
    wiki_translation = WikiTranslation.objects.get(source_block_data__data_type=key, target_block=course_block)
except WikiTranslation.DoesNotExist:
    pass
```